### PR TITLE
KAFKA-3689 - process only distinct connectionIds within processDisconnected()

### PR DESCRIPTION
**Pull request by rnpridgeon**
@ijuma discovered a possible scenario in which connectionQuotas may be [doubly-decremented](https://issues.apache.org/jira/browse/KAFKA-3689?focusedCommentId=15385962&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15385962)

This PR intends to remove that possibility by first filtering out any duplicate connectionIds. This way each

### CODE REVIEW

File: core/src/main/scala/kafka/network/SocketServer.scala [hunk 1]
```diff
@@ -512,7 +512,11 @@ private[kafka] class Processor(val id: Int,
   }
 
   private def processDisconnected() {
-    selector.disconnected.asScala.foreach { connectionId =>
+    // Duplicate connectionId entries could lead to a double decrement of the connection quota which is keyed by host address
+    val distinct = selector.disconnected.asScala.distinct
+    if (distinct.size != selector.disconnected.size)
+      warn("disconnected list has duplicates")
```
Comment (ijuma) [inline core/src/main/scala/kafka/network/SocketServer.scala]: It may be worth including the duplicates in the warning.

Comment (rnpridgeon) [inline core/src/main/scala/kafka/network/SocketServer.scala]: good point, done

**Comment by ijuma** [comment]
Thanks for the PR. A few comments:
1. It would be great to describe a scenario where we could end up with duplicate ids in the disconnected list (a test would be even better).
2. Depending on the scenario for `1`, it may be better for `Selector.disconnected` to be a `Set` instead of `List`.

I'd be interested to hear @junrao's thoughts on this.

**Comment by rnpridgeon** [comment]
Thanks @ijuma 
1. I believe this can happen when quickly killing then restarting clients. This could put connectionID into the disconnected list as failedSends once at the end of a poll(). If the same connectionID is added to the disconnected list again then we could see the same id twice. 
2. True, that is a much better solution, I'll try to get that to you today. 

I have requested a sit down with Jun to review my observations on this 1657. I'll update both PRs with my notes on that

**Comment by rnpridgeon** [comment]
closing as I have not had time to revisit.

**Pull request closed without merging**